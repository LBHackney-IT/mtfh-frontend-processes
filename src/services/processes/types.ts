export enum Trigger {
  CloseProcess = "CloseProcess",
  CancelProcess = "CancelProcess",
}
