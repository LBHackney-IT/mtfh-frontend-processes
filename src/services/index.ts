export { default as locale } from "./locale";
export { default as menu } from "./menu";
export * as processes from "./processes";
export * from "./config";
