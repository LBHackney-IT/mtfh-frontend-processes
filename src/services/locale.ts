import reactHtmlParser from "react-html-parser";

import { AssetAddress } from "@mtfh/common/lib/api/asset/v1";

const locale = {
  backButton: "Back",
  bookAppointment: "Book Appointment",
  next: "Next",
  cancel: "Cancel",
  cancelProcess: "Cancel process",
  change: "Change",
  views: {
    soleToJoint: {
      title: "Sole tenant requests a joint tenure",
      steps: {
        selectTenant: "Select person",
        checkEligibility: "Eligibility Checks",
        supportingDocuments: "Supporting documents",
        finish: "Finish",
        breachOfTenancy: "Breach of tenure check",
        requestDocuments: "Request Documents",
        reviewDocuments: "Review Documents",
        submitCase: "Submit case",
      },
      actions: {
        reassignCase: "Reassign case",
        cancelProcess: "Cancel process",
        caseActivityHistory: "Case activity history",
      },
    },
    selectTenants: {
      selectTenantHint:
        "This person will be asked for proof of relationship e.g. marriage or civil partnership certificate.",
      selectTenantLabel: "Who do you want to add as a joint tenant?",
      addToTenureText: "If the person you want to add is not listed you must first",
      addToTenureLink: "add them to the tenure",
      noHouseholdMembersOver18:
        "There are no household members over the age of 18 associated to this.",
    },
    processesMenu: {
      selectProcessDescription: () =>
        reactHtmlParser(
          "<p>Start a process by selecting a process category below. <br /> Where a process is not yet supported in <strong>Manage My Home</strong> these links will open the appropriate interim Google form.</p>",
        ),
    },
    checkEligibility: {
      autoCheckIntro:
        "Manage My Home has used tenure, person and property information to automatically check eligibility for this application.",
      passedChecks: "Passed automatic eligibilty checks",
      failedChecks: "Not eligible for a sole to joint tenure",
    },
    breachOfTenancy: {
      autoCheckIntro:
        "Manage My Home has used tenure, person and property information to automatically check eligibility for this application.",
      passedChecks: "Eligibility checks passed",
      failedChecks: "Not eligible for a sole to joint tenure",
    },
    reviewDocuments: {
      passedChecks: "Eligibility checks passed",
      documentsRequested: "Documents requested",
      checkSupportingDocumentsAppointment:
        "I need to make an appointment to check supporting documents",
      appointmentScheduled: "Office appointment scheduled",
      seenPhotographicId:
        "I confirm that an outcome letter has been sent to the resident",
      seenSecondId:
        "I confirm I have seen a second form of ID (does not have to be photographic)",
      isNotInImmigrationControl:
        "I confirm that the prospective tenant is not subject to immigration control under the Asylum And Immigration Act 1996",
      seenProofOfRelationship:
        "I confirm I have seen proof of relationship to the existing tenant",
      incomingTenantLivingInProperty:
        "I confirm I have seen 3 separate documents proving the proposed tenant has been living at the property for a minimum of 12 months",
    },
  },
  errors: {
    unableToFetchRecord: "There was a problem retrieving the record",
    unableToFetchRecordDescription:
      "Please try again. If the issue persists, please contact support.",
    unableToFindProcess: "Invalid Process",
    unableToFindProcessDescription: "We have been unable to find a process of this type.",
    unableToFindState: "Invalid State",
    unableToFindStateDescription: "We have been unable to process a state of this type.",
  },
  title: "Processes",
  loadingText: "Submitting...",
  components: {
    startProcess: {
      buttonLabel: "Start process",
      thirdPartyHeading: "Sharing information with third parties",
    },
    entitySummary: {
      tenurePaymentRef: "Tenure payment ref",
      address: (assetAddress: AssetAddress): string => {
        const {
          postPreamble,
          addressLine1,
          addressLine2,
          addressLine3,
          addressLine4,
          postCode,
        } = assetAddress;

        return [
          postPreamble,
          addressLine1,
          addressLine2,
          addressLine3,

          addressLine4,
          postCode,
        ]
          .filter((addressLine) => !!addressLine)
          .join(" ");
      },
    },
  },
};

export default locale;
