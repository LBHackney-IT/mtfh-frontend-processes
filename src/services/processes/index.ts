export * from "./sole-to-joint";
export * from "./change-of-name";
