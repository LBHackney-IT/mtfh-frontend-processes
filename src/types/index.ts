export type TargetType = "tenure" | "property" | "person";

export interface IStartProcess {
  thirdPartyCondition?: string;
  thirdPartyComponent?: React.FC;
  subHeading?: string;
  subComponent?: React.FC;
}

export interface IProcess {
  processName: string;
  title: string;
  targetType: TargetType;
  startProcess: IStartProcess;
  states: {
    [state: string]: {
      state: string;
      triggers: Record<string, string>;
    };
  };
}
