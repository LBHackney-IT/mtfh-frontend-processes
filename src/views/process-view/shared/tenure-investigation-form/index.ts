export * from "./tenure-investigation-form";
