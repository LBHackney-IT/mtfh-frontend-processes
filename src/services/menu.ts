import * as processes from "./processes";

import { Asset } from "@mtfh/common/lib/api/asset/v1/types";
import { Tenure } from "@mtfh/common/lib/api/tenure/v1/types";

interface MenuProps {
  label: string;
  link: string;
  processes?: { label: string; link: string; targetType: string }[];
  getPplQuery?: (context: any) => string;
}

const getProcessLink = (processName: string) => {
  const process = processes[processName];
  return {
    targetType: process.targetType,
    label: process.title,
    link: `/processes/${process.processName}/start`,
  };
};

const getPplQueryForRequestALegalReferral = ({ tenure, asset }): string => {
  if (!tenure || !asset) {
    return "";
  }
  const {
    householdMembers,
    tenuredAsset: { propertyReference },
  } = tenure as Tenure;
  const {
    assetAddress: { addressLine1, postCode },
  } = asset as Asset;
  const tenantNames = householdMembers
    .filter((member) => member.isResponsible)
    .map((member) => member.fullName)
    .join(", ");
  const address = addressLine1;
  const query = `usp=pp_url&entry.1699199605=${tenantNames}&entry.2086190845=${address}&entry.1956036523=${postCode}&entry.1140527432=${propertyReference}`;
  const encodedQuery = encodeURI(query);
  return encodedQuery;
};

const getPplQueryTenancyChange = ({ tenure }): string => {
  if (!tenure) {
    return "";
  }
  const {
    householdMembers,
    paymentReference,
    tenuredAsset: { fullAddress, propertyReference },
  } = tenure as Tenure;

  const tenant = householdMembers.find((member) => member.isResponsible);
  const query = `usp=pp_url&entry.1942863541=${tenant?.fullName}&entry.1090481808=${fullAddress}&entry.321634337=${propertyReference}&entry.582205322=${paymentReference}`;
  const encodedQuery = encodeURI(query);
  return encodedQuery;
};

const menu: MenuProps[] = [
  {
    label: "Sole to Joint",
    link: "https://docs.google.com/forms/d/e/1FAIpQLSdgJ9DSgGI0Aj7GO1bzLbbrArPabjS8DQwmvwb9ltB-qYYESA/viewform",
    processes: [
      getProcessLink(processes.soletojoint.processName),
      getProcessLink(processes.changeOfName.processName),
    ],
  },
  {
    label: "Other tenancy changes",
    link: "https://docs.google.com/forms/d/e/1FAIpQLSdgJ9DSgGI0Aj7GO1bzLbbrArPabjS8DQwmvwb9ltB-qYYESA/viewform",
    getPplQuery: getPplQueryTenancyChange,
  },
  {
    label: "Record a tenancy / audit homecheck",
    link: "https://docs.google.com/forms/d/e/1FAIpQLSfyXscVvgCBhbBBe7ADpV6cUqJoKrrTbU0YYRB_uXS2J4AHMQ/viewform",
  },
  {
    label: "Find legal referral response",
    link: "https://docs.google.com/spreadsheets/d/1fd6FwBXvBo2DmX8qZaXUr7hX9sE6Zr7HZsPlA_OO8Hs/edit#gid=1251146140",
  },
  {
    label: "Request a legal referral",
    link: "https://docs.google.com/forms/d/e/1FAIpQLScpHsoJQgq112HIMbW06YAFpggijwQuODJM483HGbX_PMfCMw/viewform",
    getPplQuery: getPplQueryForRequestALegalReferral,
  },
  {
    label: "Record notice served",
    link: "https://docs.google.com/forms/d/e/1FAIpQLSfRSMia4FeT6WPjrmRlsLpOXyJ4tFs3Sv_O0toxUbv_FmIjKw/viewform",
  },
  {
    label: "Rehousing",
    link: "https://docs.google.com/forms/d/e/1FAIpQLScPXSs4-xxTH2N5dUwtfmEl5mpu7kxZ6mHe86Qy_rlDNu-GbA/viewform",
  },
  {
    label: "Person",
    link: "https://docs.google.com/forms/d/e/1FAIpQLScxrWFLF7M7A3B1yzY-LZFXkV0P0qv6igcQ4wEgWHQk8DaUCw/viewform",
  },
  {
    label: "Report Japanese knotweed",
    link: "https://docs.google.com/forms/d/e/1FAIpQLSceU2H-VWi2ewaEiOAkOWULVFU2DoPSen8SRPyJrmQtRpEhfQ/viewform?gxids=7628",
  },
  {
    label: "Report overgrown tree",
    link: "https://docs.google.com/forms/d/e/1FAIpQLSfLmlLraDyOfqYhdjkOfkNWWlHUUep05G5n0GN-ewizN0Fbuw/viewform?gxids=7628",
  },
  {
    label: "Report overgrown garden",
    link: "https://docs.google.com/forms/d/e/1FAIpQLSc3mOUUEQhdRiDsS0k9XDWlfGJG0CEFAC2rnC1sYmM-idt-Nw/viewform?gxids=7628",
  },
];

export default menu;
