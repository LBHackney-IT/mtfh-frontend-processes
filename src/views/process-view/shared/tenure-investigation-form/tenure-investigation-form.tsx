import { Form, Formik } from "formik";

import {
  TenureInvestigationFormData,
  tenureInvestigationSchema,
} from "../../../../schemas/tenure-investigation";
import { locale } from "../../../../services";
import { Trigger } from "../../../../services/processes/types";
import { ProcessComponentProps, Recommendation } from "../../../../types";

import { editProcess } from "@mtfh/common/lib/api/process/v1";
import {
  Button,
  Center,
  Checkbox,
  Field,
  FormGroup,
  InlineField,
  Radio,
  RadioGroup,
  Spinner,
} from "@mtfh/common/lib/components";
import { useErrorCodes } from "@mtfh/common/lib/hooks";

const { views } = locale;

export const TenureInvestigationForm = ({
  process,
  mutate,
  setGlobalError,
}: ProcessComponentProps): JSX.Element => {
  const errorMessages = useErrorCodes();

  if (!errorMessages) {
    return (
      <Center>
        <Spinner />
      </Center>
    );
  }

  return (
    <Formik<TenureInvestigationFormData>
      initialValues={{
        tenureInvestigationRecommendation: "",
        completed: false,
      }}
      validateOnBlur={false}
      validateOnChange={false}
      validationSchema={tenureInvestigationSchema(errorMessages)}
      onSubmit={async (values) => {
        try {
          await editProcess({
            id: process.id,
            processName: process?.processName,
            etag: process.etag || "",
            processTrigger: Trigger.TenureInvestigation,
            formData: {
              tenureInvestigationRecommendation:
                values.tenureInvestigationRecommendation.toLowerCase(),
            },
            documents: [],
          });
          mutate();
        } catch (e: any) {
          setGlobalError(e.response?.status || 500);
        }
      }}
    >
      {({ dirty, errors }) => (
        <Form noValidate style={{ marginTop: 65 }}>
          <Field
            id="tenure-investigation-recommendation"
            name="tenureInvestigationRecommendation"
            label={views.tenureInvestigation.recommendedOutcome}
            type="radio"
            required
          >
            <RadioGroup>
              <Radio
                id="tenure-investigation-recommendation-approve"
                value={Recommendation.Approve}
              >
                {views.tenureInvestigation.approve}
              </Radio>
              <Radio
                id="tenure-investigation-recommendation-appointment"
                value={Recommendation.Appointment}
              >
                {views.tenureInvestigation.appointment}
              </Radio>
              <Radio
                id="tenure-investigation-recommendation-decline"
                value={Recommendation.Decline}
              >
                {views.tenureInvestigation.decline}
              </Radio>
            </RadioGroup>
          </Field>

          <FormGroup id="tenure-investigation-completed" error={errors.completed}>
            <InlineField name="completed" type="checkbox">
              <Checkbox id="tenure-investigation-completed-checkbox">
                {views.tenureInvestigation.tenureInvestigationCompleted}
              </Checkbox>
            </InlineField>
          </FormGroup>

          <Button type="submit" style={{ marginTop: 65, width: 222 }} disabled={!dirty}>
            {locale.confirm}
          </Button>
        </Form>
      )}
    </Formik>
  );
};
