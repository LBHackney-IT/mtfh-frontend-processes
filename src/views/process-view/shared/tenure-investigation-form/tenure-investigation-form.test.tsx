import { mockProcessV1, patchProcessV1, render, server } from "@hackney/mtfh-test-utils";
import { screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { locale } from "../../../../services";
import { Trigger } from "../../../../services/processes/types";
import { Recommendation } from "../../../../types";
import { TenureInvestigationForm } from "./tenure-investigation-form";

import * as processV1 from "@mtfh/common/lib/api/process/v1/service";
import * as errorMessages from "@mtfh/common/lib/hooks/use-error-codes";

const options = {
  url: "/processes/changeofname/e63e68c7-84b0-3a48-b450-896e2c3d7735",
  path: "/processes/:processName/:processId",
};

let setGlobalErrorStub;
let mutate;

describe("tenure-investigation-form", () => {
  beforeEach(() => {
    jest.resetModules();
    setGlobalErrorStub = jest.fn();
    mutate = jest.fn();
  });

  test("it renders TenureInvestigation view correctly", async () => {
    const editProcessSpy = jest.spyOn(processV1, "editProcess");
    jest.spyOn(errorMessages, "useErrorCodes").mockReturnValue({});
    const { container } = render(
      <TenureInvestigationForm
        process={{
          ...mockProcessV1,
          currentState: { ...mockProcessV1.currentState, state: "ApplicationSubmitted" },
        }}
        mutate={mutate}
      />,
      options,
    );
    await userEvent.click(screen.getByLabelText(Recommendation.Appointment));
    await userEvent.click(
      screen.getByLabelText(
        locale.views.tenureInvestigation.tenureInvestigationCompleted,
      ),
    );
    await userEvent.click(screen.getByText(locale.confirm));
    expect(editProcessSpy).toHaveBeenCalledWith(
      expect.objectContaining({
        processTrigger: Trigger.TenureInvestigation,
        formData: {
          tenureInvestigationRecommendation: Recommendation.Appointment.toLowerCase(),
        },
      }),
    );
    expect(container).toMatchSnapshot();
  });

  test("it renders error if submit fails", async () => {
    server.use(patchProcessV1(null, 500));
    render(
      <TenureInvestigationForm
        process={{
          ...mockProcessV1,
          currentState: { ...mockProcessV1.currentState, state: "ApplicationSubmitted" },
        }}
        mutate={mutate}
        setGlobalError={setGlobalErrorStub}
      />,
      options,
    );
    await expect(
      screen.findByText(Recommendation.Appointment),
    ).resolves.toBeInTheDocument();
    await userEvent.click(screen.getByLabelText(Recommendation.Appointment));
    await userEvent.click(
      screen.getByLabelText(
        locale.views.tenureInvestigation.tenureInvestigationCompleted,
      ),
    );
    await userEvent.click(screen.getByText(locale.confirm));

    await waitFor(() => expect(setGlobalErrorStub).toHaveBeenCalled());

    expect(setGlobalErrorStub).toBeCalledWith(500);
  });
});
