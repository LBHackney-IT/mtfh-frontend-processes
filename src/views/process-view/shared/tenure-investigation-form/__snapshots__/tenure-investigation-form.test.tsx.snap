// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tenure-investigation-form it renders TenureInvestigation view correctly 1`] = `
<div>
  <form
    action="#"
    novalidate=""
    style="margin-top: 65px;"
  >
    <fieldset
      class="govuk-form-group lbh-form-group"
      id="tenure-investigation-recommendation"
    >
      <legend
        class="govuk-label lbh-label"
        for="tenure-investigation-recommendation-field"
      >
        What is the recommended outcome?
        <sup
          aria-hidden="true"
        >
          *
        </sup>
      </legend>
      <div
        class="govuk-radios lbh-radios"
        id="tenure-investigation-recommendation-field"
      >
        <div
          class="govuk-radios__item"
        >
          <input
            class="govuk-radios__input"
            id="tenure-investigation-recommendation-approve"
            name="tenureInvestigationRecommendation"
            required=""
            type="radio"
            value="Approve"
          />
          <label
            class="govuk-label govuk-radios__label"
            for="tenure-investigation-recommendation-approve"
          >
            Approve
          </label>
        </div>
        <div
          class="govuk-radios__item"
        >
          <input
            class="govuk-radios__input"
            id="tenure-investigation-recommendation-appointment"
            name="tenureInvestigationRecommendation"
            required=""
            type="radio"
            value="Appointment"
          />
          <label
            class="govuk-label govuk-radios__label"
            for="tenure-investigation-recommendation-appointment"
          >
            Appointment
          </label>
        </div>
        <div
          class="govuk-radios__item"
        >
          <input
            class="govuk-radios__input"
            id="tenure-investigation-recommendation-decline"
            name="tenureInvestigationRecommendation"
            required=""
            type="radio"
            value="Decline"
          />
          <label
            class="govuk-label govuk-radios__label"
            for="tenure-investigation-recommendation-decline"
          >
            Decline
          </label>
        </div>
      </div>
    </fieldset>
    <div
      class="govuk-form-group lbh-form-group"
      id="tenure-investigation-completed"
    >
      <label
        class="govuk-label lbh-label"
        for="tenure-investigation-completed-field"
      >
        
      </label>
      <div
        class="govuk-checkboxes__item"
      >
        <input
          class="govuk-checkboxes__input"
          id="tenure-investigation-completed-field"
          name="completed"
          type="checkbox"
          value="true"
        />
        <label
          class="govuk-label govuk-checkboxes__label"
          for="tenure-investigation-completed-field"
        >
          I confirm that the tenure investigation has been completed
        </label>
      </div>
    </div>
    <button
      class="govuk-button lbh-button"
      style="margin-top: 65px; width: 222px;"
      type="submit"
    >
      Confirm
    </button>
  </form>
</div>
`;
